cmake/gdk_toolchain.cmake

```diff
@@ -0,0 +1,92 @@
+# Options to configure the toolchain.
+set(GDK_EDITION 210602 CACHE STRING "GDK edition.")
+set(GDK_PLATFORM Scarlett CACHE STRING "GDK target platform.")
+
+# Required to propagate variables when CMake calls try_compile() to test the toolchain.
+# https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_PLATFORM_VARIABLES.html
+set(CMAKE_TRY_COMPILE_PLATFORM_VARIABLES GDK_EDITION GDK_PLATFORM)
+
+cmake_path(SET gdk_gxdk_path $ENV{GameDK}/${GDK_EDITION}/GXDK NORMALIZE)
+
+# Set C/C++ compile flags and additional include directories.
+foreach(lang C CXX)
+    set(CMAKE_${lang}_FLAGS_INIT "")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /D_GAMING_XBOX")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /DWINAPI_FAMILY=WINAPI_FAMILY_GAMES")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /D_ATL_NO_DEFAULT_LIBS")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /D__WRL_NO_DEFAULT_LIB__")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /D__WRL_CLASSIC_COM_STRICT__")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /D_CRT_USE_WINAPI_PARTITION_APP")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /DWIN32_LEAN_AND_MEAN")
+    string(APPEND CMAKE_${lang}_FLAGS_INIT " /favor:AMD64")
+
+    if(GDK_PLATFORM STREQUAL Scarlett)
+        string(APPEND CMAKE_${lang}_FLAGS_INIT " /D_GAMING_XBOX_SCARLETT")
+        string(APPEND CMAKE_${lang}_FLAGS_INIT " /arch:AVX2")
+    elseif(GDK_PLATFORM STREQUAL XboxOne)
+        string(APPEND CMAKE_${lang}_FLAGS_INIT " /D_GAMING_XBOX_XBOXONE")
+        string(APPEND CMAKE_${lang}_FLAGS_INIT " /arch:AVX")
+    endif()
+
+    set(CMAKE_${lang}_STANDARD_INCLUDE_DIRECTORIES ${gdk_gxdk_path}/gameKit/Include/${GDK_PLATFORM})
+
+    set(CMAKE_${lang}_STANDARD_LIBRARIES "onecoreuap_apiset.lib" CACHE STRING "" FORCE)
```

Comment:
Why don't use a toolchain file like https://github.com/microsoft/onnxruntime/blob/master/cmake/wcos_rules_override.cmake

Reply:
This is a CMake toolchain file. I assume you mean setting `CMAKE_USER_MAKE_RULES_OVERRIDE=wcos_rules_override.cmake`. I wasn't even aware of this variable until you pointed it out, but this seems to work as well.

I know force-setting cache variables is generally frowned upon -- it prevents users from overriding values on the command line -- but the [`CMAKE_<LANG>_STANDARD_LIBRARIES`](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_STANDARD_LIBRARIES.html) docs say this variable is _only_ to be set in toolchain files or CMake's platform modules.

I'm not really sure which approach is best practice here. Clarify what you prefer and I'll be happy to make the change!

Reply:
We have onecore build. The binaries in our github release page are indeed onecore binaries.  As you pointed out, they were built with CMAKE_USER_MAKE_RULES_OVERRIDE=wcos_rules_override.cmake. Our build.py has a "--enable_wcos" parameter, it is for this purpose. But, the build option does more than just setting two variables.  Our cmake file has code like 
```
if(CMAKE_CXX_STANDARD_LIBRARIES MATCHES kernel32.lib)
```
in a few places. So basically we assume we will only have two build types: desktop build and onecore build. And onecore build should use onecore CRT. So, when you add the GDK build option in, it automatically falls into the onecore build category. Please search "CMAKE_CXX_STANDARD_LIBRARIES" in our code and be aware of the differences. You didn't add the "if...else", but they will impact your build. I hope you can reuse the "CMAKE_USER_MAKE_RULES_OVERRIDE=wcos_rules_override.cmake" usage.

Reply:
Thanks. I'm already setting `CMAKE_TOOLCHAIN_FILE=gdk_toolchain.cmake` in build.py when `--use_gdk` is given, and this is giving me the results I want; using the onecore CRT is *desired* when building with the GDK. I double-checked all references to`if(CMAKE_CXX_STANDARD_LIBRARIES MATCHES kernel32.lib)` and they look fine to me.

I think it's probably best to keep this toolchain as-is rather than mixing both gdk_toolchain.cmake and wcos_rules_override.cmake: at some point it might be desirable to link xgameplatform.lib instead of onecoreuap_apiset.lib.